File: src/main/java/store/wckd/server/service/UserService.java
package store.wckd.server.service;

import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import store.wckd.server.dto.UserCreateDTO;
import store.wckd.server.dto.UserUpdateDTO;
import store.wckd.server.entity.User;
import store.wckd.server.repository.UserRepository;

@Service
public class UserService {

    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public Flux<User> findAll(int page) {
        return Flux.fromIterable(userRepository.findAll());
    }

    public Mono<User> create(UserCreateDTO userCreateDTO) {
        User user = new User(
                userCreateDTO.getUsername(),
                userCreateDTO.getEmail(),
                userCreateDTO.getPassword()
        );

        return Mono.just(userRepository.save(user));
    }

    public Mono<User> findByUsername(String username) {
        return Mono.just(userRepository.findByUsername(username));
    }

    public Mono<User> findById(long id) {
        return Mono.justOrEmpty(userRepository.findById(id));
    }

    public Mono<User> updateById(long id, UserUpdateDTO userUpdateDTO) {
        User user = new User(
                id,
                userUpdateDTO.getUsername(),
                userUpdateDTO.getEmail(),
                userUpdateDTO.getPassword()
        );

        return Mono.just(userRepository.save(user));
    }

    public Mono<Void> deleteById(long id) {
        userRepository.deleteById(id);

        return Mono.empty();
    }

}

File: src/main/java/store/wckd/server/service/UserService.kt
package store.wckd.server.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import store.wckd.server.dto.user.UserCreateDTO
import store.wckd.server.dto.user.UserUpdateDTO
import store.wckd.server.entity.User
import store.wckd.server.exception.EntityNotFoundException
import store.wckd.server.repository.UserRepository

@Service
class UserService(private val userRepository: UserRepository) {
    suspend fun findAll(page: Int): Page<User> {
        return userRepository.findAll(PageRequest.of(page, 15))
    }

    suspend fun create(userCreateDTO: UserCreateDTO): User {
        val user = User(
                username = userCreateDTO.username,
                email = userCreateDTO.email,
                password = userCreateDTO.password
        )

        return userRepository.save(user)
    }

    suspend fun findByUsername(username: String): User {
        return userRepository.findByUsername(username).orElseThrow(::EntityNotFoundException)
    }

    suspend fun findById(id: Long): User {
        return userRepository.findById(id).orElseThrow(::EntityNotFoundException)
    }

    suspend fun updateById(id: Long, userUpdateDTO: UserUpdateDTO): User {
        val user = User(
                id,
                userUpdateDTO.username,
                userUpdateDTO.email,
                userUpdateDTO.password
        )

        return userRepository.save(user)
    }

    suspend fun deleteById(id: Long) {
        userRepository.deleteById(id)
    }
}